### DISCUSSION TITLE
Design of read query - filter, join, fields, sort, pagination

### DISCUSSION THREAD

**Discussion by kdhrubo** [9tigerio/db2rest#245] [Ideas]
### Current Syntax

`<url_base>/{table/view}?select=<list_of_columns>&filter=<rsql_filter_statement>page=<page_size>&size=<page_size>&sort=<col_id>&sort=<col_name>,<sort_order>`

- **url_base** : server base URL.
- **table/view** : root table or view.
- **select**: list of columns to be retrieved from the query execution. If the list is empty or the `select` request parameter is missing, then all the columns of the root table/view are retrieved. `select` can also specify implicit joins or resource embedding. 

`select=film_id:id,title,description,release_year:yearOfRelease,language_id;language(name)
`
In this case, DB2Rest checks if an FK relationship exists between film -> language tables. If a foreign key exits, then an implicit join is applied and data is retrieved. The retrieved rows - include columns from the root table and the field `name` from the `language` table.  

This is a very concise syntax. However, it is difficult to understand by looking at it. Also, it results in complex parser code. In case, there is no foreign key between the tables, there is no way currently to specify an explicit join. 

- **filter** : This results in the where clause of the SQL statement. This follows the RSQL specification. However, the RSQL specification only supports the columns of the root table. It also needs to support fields of any join tables going forward. 

- **page** : page number of the search result to retrieve. 
- **size** : specifies the number of records to retrieve

`page` and `size` are used to compute `offset` and `limit` for offset pagination. 

Is it possible to combine `page` and `size` along with `filter` for more efficient `seek` pagination? 

- **sort**: parameter list to specify the fields to sort and order. This one needs no change, is verbose, and easy to understand. 

### Proposal for new Syntax and defaults 

1. Rename the `select` request parameter to `fields` 
2. `fields` to include fields of both root and join table columns. (what if a table has two foreign keys to another table - viz. `film` has two fields `language_id` fields as foreign keys of the `language` table - then how this will be specified?. See below)
3. Introduce a new request parameter `join` or `embed`. **There can be zero or more `join` parameters.** 
4. `join=language`: looks for an implicit join relation between the root table (e.g.: `film`) and the `language` table. This will apply inner join if the foreign key relationship is not nullable, otherwise left join is applied. Also, this will include all the columns from the `language` table as part of the columns list. 
5. `join=language?fields:language_id,name`: Does same as in item#4, however only includes the columns `language_id`, `name`. 
6. `join=language?fields:language_id,name&condition:language_id, language_id; inner`: In this case, explicit join is applied because the foreign keys are not defined. The columns `language_id`, and `name` are included. There is an explicit inner-join between the `language_id` of the root table (e.g. `film`) and the `language.language_id` column.   
7. `join=language?fields:language_id,name&condition:alternate_language_id, language_id; left`: In this case, there is an explicit left-join between the `alternate_language_id` of the root table (e.g. `film`) and the `language.language_id` column.   

In the case of no#6, no#7 - the first column belongs to the root table and the second column belongs to the non-root table. In case, the first table is non-root, use the `<table_name>.<column_name>` syntax. 

If no column from the join table is to be included, specify that with `-`: e.g `join=language?fields=-`

In case of implicit join and foreign key is not found - raise the error `unable to find table relationship`. 

Do not include any table alias only column alias are allowed. 

8. The `schema` request parameter is to be removed completely. Schema/catalog to be detected based on access permission when looking up a table in the metadata cache. 
9. `filter` - can include conditions for filtering based on join table columns - e.g 'year_of_release=eq=2023;director.country_id=eq=1' this will translate to 

```
where 
       film.year_of_release = 2003
and director.country_id = 1
```

10. Seek pagination can be done with only size and filter conditions (thoughts ?)
11. Subquery - **not supported** - please use the custom query execution option. 

That should be it. Please let me know your thoughts if anything I have overlooked or missed.

**Comment by thadguidry** [comment]
I feel like mapping or key/value construction should either use `=` or `:` , but your examples use `;` and that feels awkward to me.  Semicolon also can legitimately be part of an identifer in PostgreSQL when the identifier is quoted:
> The semicolon (;) terminates an SQL command. It cannot appear anywhere within a command, except within a string constant or quoted identifier.
https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

You'll need to account for usage of quoted identifiers somehow.

Let's breakdown this:
`&condition:alternate_language_id, language_id; left`
It can be subtraced as:
`&condition:alternate_language_id; left`
`&condition:language_id; left`  <-- this guy feels weird and also "language_id;" itself might actually be an identifier (including the semicolon)...so...

More readable?
`&condition(alternate_language_id, language_id AND left)` <-- RSQL actual parse where semicolon is AND shortcut construct, but "AND left" seems it's lacking context of the join condition type and brings semantic ambiguity. 

More readable and more context and allows for unicode and also multi-conditions in future?
`&condition=(left:"alternate_language_id, language_id" right:"data_id" right:"U&"d!0061t!+000061" UESCAPE '!'")` <-- but problem there on the `right:` join condition with the double quote and parser eagerness... so maybe ... dunno, ya'll need to research the naming conventions doc link above and come up with something.  How to account for identifiers with escaping? (my eyes 👀 need a break)

**Reply by kdhrubo** [reply]
This can certainly be worked upon.

So this join part is where we need to think a bit more.

How about the other components ?

**Reply by kdhrubo** [reply]
Just like sort, allow multiple `joins` parameters

Example1: 

`join=language?fields=language_id,name|condition='alternate_language_id=LEFT=language_id`

& 

`join=language?fields=language_id,name|condition='language_id=INNER=language_id;language_id=gt=10`



How about this for multi-join?

**Reply by susanta-mukhopadhyay** [reply]
Above is Ok, but in my opinion end user may be confused with lots of '=' used in the syntax for larger queries. I have basic idea about FIQL. If we can introduce symbols which can be parsed easily, we may also think of following options for better readability:

**Option 1:**
join=language?fields=(language_id,name)|condition=('language_id=INNER=language_id;language_id=gt=10)

**Option 2:**
join=language?fields:language_id,name|condition:'language_id=INNER=language_id;language_id=gt=10

**Reply by thadguidry** [reply]
what does `|` before `condition` mean?
kinda agree with @susanta-mukhopadhyay that the RSQL `=INNER=` can possibly be confusing.  It's concise syntax, don't get me wrong, but I don't think you are addressing the fact that...
1. There are unquoted identifiers - fine, great.
2. But a database could also have and use quoted identifiers - <-- THINK and SHOW how this will work in the query URL syntax.

> Quoted identifiers can contain any character, except the character with code zero.
Which means the below can happen in reality where even `=` or any char! except `char(0)` could be part of a table or field identifier (name).

`join=language?fields=language_id,name|condition='_lang=uage=id=INNER=langua=ge_i=d;language_id=gt=10`

We MUST account for quoted SQL identifiers (table identifiers, column identifiers, object identifiers, etc.) which can contain any characters, including Unicode chars including CJK multibyte chars.  Database server encoding is UTF-8 nowadays by default.

You need to add double quote handling - i.e. maybe make it a rule that in the URL query syntax that all identifiers must be wrapped in double quotes. (as we did with the Freebase API - which turned into Google API's) somehow, someway.
That's what I was trying to tell you and show you in my last paragraph prior comment.  Did you read the PostgreSQL Docs link in my comment?  Please do.

**Reply by kdhrubo** [reply]
separator. I think this can be simplified bit further? merge the fields with table name with ()

**Reply by thadguidry** [reply]
The other thing that I don't see addressed is when a [join](https://www.postgresql.org/docs/16/tutorial-join.html) has multiple keys joined...i.e. multiple `ON` statements.
Example:
```sql
SELECT w1.city, w1.temp_lo AS low, w1.temp_hi AS high,
       w2.city, w2.temp_lo AS low, w2.temp_hi AS high
    FROM weather w1 JOIN weather w2
        ON w1.temp_lo < w2.temp_lo AND w1.temp_hi > w2.temp_hi;
```

**Reply by thadguidry** [reply]
BTW, I found out today that Groovy has a GQuery DSL https://groovy-lang.org/using-ginq.html

**Reply by kdhrubo** [reply]
This works on collection objects. Won't switch from mybatis dynamic.

**Comment by kdhrubo** [comment]
```

SELECT w1.city, w1.temp_lo AS low, w1.temp_hi AS high,
       w2.city, w2.temp_lo AS low, w2.temp_hi AS high
    FROM weather w1 JOIN weather w2
        ON w1.temp_lo < w2.temp_lo AND w1.temp_hi > w2.temp_hi;
```


```
/weather?fields=city, temp_lo:low, temp_high:high
&join=weather?joinFields=city,temp_low:low,temp_hi:high&joinFilter=temp_lo=lt=temp_lo;temp_hi=le=temp_hi
```

So I am trying to keep this similar to the root table with `fields`, `filter` with `joinFields`, `joinFilter` for `joinTable`

**Restriction** - is you have to put this in order, db2rest will walk from root --> child 1 .... child n

Please share more of these scenarios so that the syntax can be firmed up.

**Comment by kdhrubo** [comment]
```
/fims?fields=id, title, year_of_release:yearOfRelease
&join=language?joinFields=language_id,name&joinFilter=language_id=eq=language_id
&join=language?joinFields=language_id,name&joinFilter=alternate_language_id=eq=language_id
```
How to indicate LEFT JOIN RIGHT JOIN?

**Reply by thadguidry** [reply]
`joinType="left"`

**Reply by kdhrubo** [reply]
Very difficult to do this with request params. 
Hence changing to a new endpoint with POST with joins specified in the JSON body.

**Comment by kdhrubo** [comment]
@susanta-mukhopadhyay @thadguidry - take some time to review the syntax of ORDS 

https://www.thatjeffsmith.com/archive/2019/09/some-query-filtering-examples-in-ords/

**Reply by thadguidry** [reply]
Thoughts on what I see on the limited ORDS syntax shown on that site:
- That's using JSON object notation.
- It is slightly less concise and closer to just the raw SQL query might be.
- SQL keywords seem to need to have a `$` prefix and identifiers do not.

**Reply by kdhrubo** [reply]
Not very good syntax IMO

**Reply by kdhrubo** [reply]
Can you please list various scenarios possible with Join with an SQL example and then I can try to map it to RSQL.

@thadguidry

**Reply by thadguidry** [reply]
## Developing with ORDS
Reading more on ORDS briefly, I came across this:
>Swagger can be used to develop OpenAPI (V2 or V3) specifications for your REST APIs. Once templates, handlers, and parameters are represented in JSON, they can be imported into ORDS as new modules. Afterward, you can supply the SQL and PL/SQL required for all of your handlers. Operating this way will greatly improve your workflow and make building Oracle REST APIs with ORDS more efficient.

Where I question that this is efficient way.  Maybe it's just me.  But I wanted to dig deeper and so read around more in their various docs sections and Best Practices documentation (where they like to hide things from my experience 😄 )

Also there is a claim that:
>Auto REST for tables and views, RPC for PL/SQL

But that requires and assumes efficient because you'd be using the [Oracle SQL Developer Web interface to design and develop with](https://www.oracle.com/database/technologies/appdev/rest/best-practices/#Modeling).  I guess that's OK.  If you use Oracle, it seems like you'll likely just use all their tooling and support (where some of it is indeed an extra package license per developer! woah!)

> - Easily upload or retrieve files to and from your tables
> - Results & responses auto formatted to {JSON}
> - Supports GET, PUT, POST, & Delete Handlers

### Request syntax rules
- [Developing Oracle REST Data Services Applications](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/developing-REST-applications.html#GUID-A1CD111F-724B-4E91-8202-FA899EE521F1)
  - 👀 [About Request Path Syntax Requirements](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/developing-REST-applications.html#GUID-8D2CA518-2F13-4503-8DBF-62C5F2C681B7) 


### AutoRest
[Automatic Enabling of Schema Objects for REST Access (AutoREST)](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/developing-REST-applications.html#GUID-4CE630AA-2F06-41D9-96F6-DA77AB1E6395)

It mentions that you still have to use SQL or PL/SQL to do joins even with AutoRest enabled:
> AutoRest lets you quickly expose data but (metaphorically) keeps you on a set of guide rails. For example, you cannot
customize the output formats or the input formats, or do extra validation.
On the other hand, manually created resource modules require you to specify the SQL and
PL/SQL to support the REST resources. Using resource modules requires more effort, but
offers more flexibility; for example, you can customize what fields are included, **do joins
across multiple tables**, and validate the incoming data **using PL/SQL**.

## Join support within ORDS specific service offerings
No Join support unless SQL used
- Looks like Oracle didn't do the extra work in ORDS Rest API where not even with [SODA](https://docs.oracle.com/en/database/oracle/simple-oracle-document-access/adsdi/overview-soda.html) (Simple Oracle Document Access - a set of NoSQL-style APIs that let you create and store collections of documents (in particular JSON) in Oracle Database) can you perform joins without knowing SQL and applying it through not the URL `/query` type but you must use SQL through one of the other query types like `[ "dml", "ddl", "plsql" , "sqlplus" , "ignore", "transaction-control", "session-control", "system-control", "jdbc", "other" ]`.

Join support directly in GraphQL however
- But Join support is also there in other ORDS forms like GraphQL but you must [enable both](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/graphql-oracle-rest-data-services.html#GUID-A0E0529B-0F47-40D9-B88D-BEE9D8CD6C3D) GraphQL and enable your objects for GraphQL as well:
    > Any REST-Enabled table or view of an Oracle REST Data Services enabled schema
can be accessed through GraphQL queries. For a REST-Enabled object to be mapped
into a GraphQL type, it is necessary that it has one or multiple primary keys associated
to the object. If this condition is not satisfied, then the ROWID pseudo column is used
to guarantee that the objects obtained in a query are unique and are not a duplicate
derived from a join.

**Reply by kdhrubo** [reply]
So we are on right track IMO with RSQL and supporting both implicit and explicit join.

**Reply by thadguidry** [reply]
Yeah, but SQL-like is unavoidable.  What I mean is... a user must give us a join type and the join conditions, but then we take care of making valid SQL for them and returning row objects

**Reply by thadguidry** [reply]
And look at that link for the Request Path Syntax Rules ... we might learn something (or avoid some things)

**Reply by thadguidry** [reply]
At this point...would be good to know what T***bo thinks?

**Reply by kdhrubo** [reply]
What is request path syntax?

**Reply by thadguidry** [reply]
In my previous long comment above
> Request syntax rules
[Developing Oracle REST Data Services Applications](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/developing-REST-applications.html#GUID-A1CD111F-724B-4E91-8202-FA899EE521F1)
👀 [About Request Path Syntax Requirements](https://docs.oracle.com/en/database/oracle/oracle-rest-data-services/23.4/orddg/developing-REST-applications.html#GUID-8D2CA518-2F13-4503-8DBF-62C5F2C681B7)

**Reply by susanta-mukhopadhyay** [reply]
I agree with @thadguidry . Request path rules are good to take care in our queries.

**Reply by kdhrubo** [reply]
Very difficult to do this with request params.
Hence changing to a new endpoint with POST with joins specified in the JSON body.

**Reply by kdhrubo** [reply]
Let the syntax evolve now @thadguidry @susanta-mukhopadhyay @rajivmca2004 @souravroy @musibs

**Comment by kdhrubo** [comment]
Opening a new discussion for the join request body syntax evolution. This one is already way too long. We kind of all convinced on filter, fields, sort, pagination, and join. (We may also consider subquery in the future :))
